File: CS/Form1.cs
using System;
using System.Data;
using System.Windows.Forms;
using DevExpress.XtraReports.UI;
// ...

namespace dxKB17175 {
    public partial class Form1 : Form {
        public Form1() {
            InitializeComponent();
        }

        private void simpleButton2_Click(object sender, EventArgs e) {
            XtraReport1 report = new XtraReport1(checkEdit1.Checked ? FillDatasetFromGrid() : nwindDataSet, gridView1);
            ReportPrintTool printTool = new ReportPrintTool(report);
            printTool.ShowPreviewDialog();
        }

        private DataSet FillDatasetFromGrid() {
            nwindDataSet ds = new nwindDataSet();
            int rowHandle;
            DataRow gridRow;
            for (int i = 0; i < gridView1.RowCount; i++) {
                rowHandle = gridView1.GetVisibleRowHandle(i);
                if (!gridView1.IsGroupRow(rowHandle)) {
                    gridRow = gridView1.GetDataRow(rowHandle);
                    ds.Tables[0].Rows.Add(gridRow.ItemArray);
                }
            }
            return ds;
        }

        private void Form1_Load(object sender, EventArgs e) {
            this.customersTableAdapter.Fill(this.nwindDataSet.Customers);
        }

        private void simpleButton1_Click(object sender, EventArgs e) {
            XtraReport1 report = new XtraReport1(checkEdit1.Checked ? FillDatasetFromGrid() : nwindDataSet, gridView1);
            ReportDesignTool designTool = new ReportDesignTool(report);
            designTool.ShowDesignerDialog();
        }
    }
}

File: VB/Form1.vb
Imports System
Imports System.Data
Imports System.Windows.Forms
Imports DevExpress.XtraReports.UI

' ...
Namespace dxKB17175

    Public Partial Class Form1
        Inherits Form

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub simpleButton2_Click(ByVal sender As Object, ByVal e As EventArgs)
            Dim report As XtraReport1 = New XtraReport1(If(checkEdit1.Checked, FillDatasetFromGrid(), nwindDataSet), gridView1)
            Dim printTool As ReportPrintTool = New ReportPrintTool(report)
            printTool.ShowPreviewDialog()
        End Sub

        Private Function FillDatasetFromGrid() As DataSet
            Dim ds As nwindDataSet = New nwindDataSet()
            Dim rowHandle As Integer
            Dim gridRow As DataRow
            For i As Integer = 0 To gridView1.RowCount - 1
                rowHandle = gridView1.GetVisibleRowHandle(i)
                If Not gridView1.IsGroupRow(rowHandle) Then
                    gridRow = gridView1.GetDataRow(rowHandle)
                    ds.Tables(CInt(0)).Rows.Add(gridRow.ItemArray)
                End If
            Next

            Return ds
        End Function

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs)
            customersTableAdapter.Fill(nwindDataSet.Customers)
        End Sub

        Private Sub simpleButton1_Click(ByVal sender As Object, ByVal e As EventArgs)
            Dim report As XtraReport1 = New XtraReport1(If(checkEdit1.Checked, FillDatasetFromGrid(), nwindDataSet), gridView1)
            Dim designTool As ReportDesignTool = New ReportDesignTool(report)
            designTool.ShowDesignerDialog()
        End Sub
    End Class
End Namespace
